src/transport/manager/mod.rs

```diff
@@ -1105,239 +795,55 @@ impl TransportManager {
+            if let PeerState::Opening {
+                connection_id,
+                transports,
+                ..
+            } = previous_state
+            {
+                // cancel all pending dials
+                transports.iter().for_each(|transport| {
+                    self.transports
+                        .get_mut(transport)
+                        .expect("transport to exist")
+                        .cancel(connection_id);
+                });
 
-                    let (record, dial_record) = match dial_record.take() {
-                        Some(mut dial_record) =>
-                            if dial_record.address() == endpoint.address() {
-                                dial_record.set_connection_id(endpoint.connection_id());
-                                (dial_record, None)
-                            } else {
-                                (
-                                    AddressRecord::new(
-                                        &peer,
-                                        endpoint.address().clone(),
-                                        SCORE_CONNECT_SUCCESS,
-                                        Some(endpoint.connection_id()),
-                                    ),
-                                    Some(dial_record),
-                                )
-                            },
-                        None => (
-                            AddressRecord::new(
-                                &peer,
-                                endpoint.address().clone(),
-                                SCORE_CONNECT_SUCCESS,
-                                Some(endpoint.connection_id()),
-                            ),
-                            None,
-                        ),
-                    };
-
-                    context.state = PeerState::Connected {
-                        record,
-                        dial_record,
-                    };
-                }
-            },
-            None => {
-                peers.insert(
-                    peer,
-                    PeerContext {
-                        state: PeerState::Connected {
-                            record: AddressRecord::new(
-                                &peer,
-                                endpoint.address().clone(),
-                                SCORE_CONNECT_SUCCESS,
-                                Some(endpoint.connection_id()),
-                            ),
-                            dial_record: None,
-                        },
-                        addresses: AddressStore::new(),
-                        secondary_connection: None,
-                    },
-                );
+                // since an inbound connection was removed, the outbound connection can be
+                // removed from pending dials
+                //
+                // TODO: This may race in the following scenario:
```

Comment:
Can't understand now, is it difficult to fix / how bad is it? Do we need an issue for this?